arch/x86/core/intel64/locore.S

```diff
@@ -0,0 +1,599 @@
+
+	/* jump into long mode, reload the segment registers (again). */
+
+	jmpl $X86_KERNEL_CS_64, $1f
+.code64
+1:	movl $X86_KERNEL_DS_64, %eax
+	movw %ax, %ds
+	movw %ax, %es
+	movw %ax, %ss
+	movw %ax, %fs
+	movw %ax, %gs
+
+	movl $X86_KERNEL_TSS, %eax
+	ltr %ax
+
+	cld
+	xorl %eax, %eax
+	movq $__bss_start, %rdi
+	movq $__bss_num_qwords, %rcx
+	rep stosq
+
+#ifdef CONFIG_X86_MULTIBOOT_INFO
+	/* If EBX isn't 0, then we have a valid multiboot info struct. */
+
+	testl %ebx, %ebx
+	jz 1f
+	movl %ebx, %esi
+
+	movl $__X86_MULTIBOOT_INFO_SIZEOF, %ecx
+	movq $x86_multiboot_info, %rdi
+	rep
+	movsb
+1:
+#endif /* CONFIG_X86_MULTIBOOT_INFO */
+
+	/*
+	 * set up SSE in case something uses the floating-point unit during
+	 * early initialization (either directly, or if GCC gets clever)
+	 */
+
+	xorl %edi, %edi
+	call x86_sse_init
+
+	/* don't replace CALL with JMP; honor the ABI stack alignment! */
+
+	call z_cstart
+
+stop:	jmp stop
+
+/*
+ * void x86_sse_init(struct k_thread *thread);
+ *
+ * Initialize floating-point state to something sane. If 'thread' is
+ * not NULL, then the resulting FP state is saved to thread->arch.sse.
+ */
+
+.global x86_sse_init
+x86_sse_init:
+	fninit
+	ldmxcsr mxcsr
+	testq %rdi, %rdi
+	jz 1f
+	fxsave _thread_offset_to_sse(%rdi)
+1:	retq
+
+mxcsr:	.long X86_MXCSR_SANE
+
+/*
+ * FIXME: The multiboot header is identical (for obvious reasons) to the
+ * version in ia32/crt0.S. They should be refactored into a common file.
+ */
+
+#ifdef CONFIG_X86_MULTIBOOT
+	.align 4
+	.long X86_MULTIBOOT_HEADER_MAGIC
+	.long X86_MULTIBOOT_HEADER_FLAGS
+	.long -(X86_MULTIBOOT_HEADER_MAGIC + X86_MULTIBOOT_HEADER_FLAGS)
+#ifdef CONFIG_X86_MULTIBOOT_FRAMEBUF
+	.fill 5,4,0				/* (unused exec layout) */
+	.long 0					/* linear graphics mode */
+	.long CONFIG_X86_MULTIBOOT_FRAMEBUF_X	/* width */
+	.long CONFIG_X86_MULTIBOOT_FRAMEBUF_Y	/* height */
+	.long 32				/* depth */
+#endif /* CONFIG_X86_MULTIBOOT_FRAMEBUF */
+#endif
+
+/*
+ * XXX: describe __swap, __resume, stacks
+ */
+
+.globl _k_neg_eagain	/* from errno.c: int _k_neg_eagain = -EAGAIN; */
+
+.globl __swap
+__swap:
+	movq $_kernel, %rsi
+	movq _kernel_offset_to_current(%rsi), %rsi
+
+	orb $_THREAD_SWAPPED, _thread_offset_to_thread_state(%rsi)
+
+	movl _k_neg_eagain, %eax
+	movl %eax, _thread_offset_to_rax(%rsi)
+	popq %rax
+	movq %rax, _thread_offset_to_rip(%rsi)
+	movq %rsp, _thread_offset_to_rsp(%rsi)
+	movl %edi, %edi /* N.B.: zero extend */
+	movq %rdi, _thread_offset_to_rflags(%rsi)
+	movq %rbx, _thread_offset_to_rbx(%rsi)
+	movq %rbp, _thread_offset_to_rbp(%rsi)
+	movq %r12, _thread_offset_to_r12(%rsi)
+	movq %r13, _thread_offset_to_r13(%rsi)
+	movq %r14, _thread_offset_to_r14(%rsi)
+	movq %r15, _thread_offset_to_r15(%rsi)
+
+	movq $(ist + CONFIG_ISR_STACK_SIZE), %rsp
+
+	/* fall through to __resume */
+
+/*
+ * Entry:
+ *   RSP = top of ist
+ */
+
+__resume:
+	movq $_kernel, %rdi
+	movq _kernel_offset_to_ready_q_cache(%rdi), %rsi
+	movq %rsi, _kernel_offset_to_current(%rdi)
```

Comment:
So... this scheme doesn't work with SMP.  There is no ready_q_cache variable possible there and you need to call into the scheduler to fetch the next thread both here and in interrupt exit.  Is there a reason you dropped the z_arch_switch() implementation that already gave you that?

Reply:
> Finally got some time to look at this closely, or at least at the core layer. Lots of notes, mostly of the form "my earlier stuff was better, please bring those bits back".

This should really be viewed as a forward-port of the x86 32-bit arch to 64-bit, not as some residual from your xuk/x86_64 work.  I didn't base this work on yours.  I didn't even really look at it once I decided it wasn't the right path forward for my use cases.

> Most of them are things that will only matter to microbenchmarks, or people trying to maintain or evolve the code long term, I guess (I mean, I honestly kind of hate your context switch code, but it clearly looks like it works). But the __swap() vs. z_arch_switch() thing is something you're going to need to fix for sure before SMP lands (er, lands again, I guess).

The requirement for `z_arch_switch()` with the SMP scheduler is an artifact of your implementation of said scheduler, not a reflection of some inherent problem with `__swap()` primitive on SMP.  I quite deliberately constructed this code with the intent that it be usable in multi-processor systems.  More on that later.

> A last thing that doesn't really fit in a line comment is that the xuk layer had a z_arch_printk_char_out() handler so that printk() would work (to either the serial port or the VGA console) from the very first 64 bit instruction. That really would be nice to have back.

The x86 32-bit doesn't have this, so neither does the 64-bit version.  Again, I didn't base my work on xuk/x86_64, I didn't take anything away that I can "give back".